`pointerup` is received before boundary events after mouse released over newly attached child element

lilyspiniolas opened the pull request:
#### e371b8a2034ad01151bcb56594a501550f44482d
<pre>
`pointerup` is received before boundary events after mouse released over newly attached child element
<a href="https://bugs.webkit.org/show_bug.cgi?id=299150">https://bugs.webkit.org/show_bug.cgi?id=299150</a>
<a href="https://rdar.apple.com/160913756">rdar://160913756</a>

Reviewed by Abrar Rahman Protyasha.

When the mouse is down over an element and a child element is attached (located
underneath the cursor) and the mouse is then released, instead of getting
`pointerout@parent, pointerover@child, pointerenter@child, pointerup@child`, we were
getting `pointerup@child, pointerout@parent, pointerover@child, pointerenter@child`.

The reason for this is that, in `EventHandler::handleMouseReleaseEvent`,
`dispatchMouseEvent` is called with parameter `FireMouseOverOut::No`, followed
by `swallowAnyClickEvent`, which itself calls `dispatchMouseEvent` with
parameter `FireMouseOverOut::Yes`. The `FireMouseOverOut` parameter is used when
calling `updateMouseEventTargetNode`, which fires boundary events when the parameter
is `FireMouseOverOut::Yes`.

To fix the order of events, simply update the initial `dispatchMouseEvent` call to use
`FireMouseOverOut::Yes` so that boundary events are fired before `pointerup`, and update
the `dispatchMouseEvent` call in `swallowAnyClickEvent` to `FireMouseOverOut::No` since
boundary events will have already been handled by the call to `dispatchMouseEvent` which
happens prior. `handleMouseDoubleClickEvent` has been updated to use the same behavior
as well for consistency.

Tested by the first and fifth asserts in
`LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended.html?mouse`

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_boundary_events_attributes_during_drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseenter-mouseleave-on-drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=16-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=42-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&amp;button=0&amp;buttons=1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=16-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=42-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::swallowAnyClickEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):

Canonical link: <a href="https://commits.webkit.org/300273@main">https://commits.webkit.org/300273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1341f06629f1f68b989c0aba0be039927f24fcaa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128488 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74028 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a5fdfa84-898e-495a-9e5f-538e7077ae73) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42352 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92667 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/11df4548-6415-4d37-9937-6c73b95cd3cb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124887 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33766 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109194 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73326 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e7be718e-6932-4809-9a20-d6e879b8956d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32774 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27361 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71991 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131259 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101228 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101096 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25643 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46467 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24586 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48731 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54465 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48201 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51551 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49881 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
> When the mouse is down over over an element

Small typo in the commit message

lilyspiniolas committed:
`pointerup` is received before boundary events after mouse released over newly attached child element
https://bugs.webkit.org/show_bug.cgi?id=299150
rdar://160913756

Reviewed by Abrar Rahman Protyasha.

When the mouse is down over an element and a child element is attached (located
underneath the cursor) and the mouse is then released, instead of getting
`pointerout@parent, pointerover@child, pointerenter@child, pointerup@child`, we were
getting `pointerup@child, pointerout@parent, pointerover@child, pointerenter@child`.

The reason for this is that, in `EventHandler::handleMouseReleaseEvent`,
`dispatchMouseEvent` is called with parameter `FireMouseOverOut::No`, followed
by `swallowAnyClickEvent`, which itself calls `dispatchMouseEvent` with
parameter `FireMouseOverOut::Yes`. The `FireMouseOverOut` parameter is used when
calling `updateMouseEventTargetNode`, which fires boundary events when the parameter
is `FireMouseOverOut::Yes`.

To fix the order of events, simply update the initial `dispatchMouseEvent` call to use
`FireMouseOverOut::Yes` so that boundary events are fired before `pointerup`, and update
the `dispatchMouseEvent` call in `swallowAnyClickEvent` to `FireMouseOverOut::No` since
boundary events will have already been handled by the call to `dispatchMouseEvent` which
happens prior. `handleMouseDoubleClickEvent` has been updated to use the same behavior
as well for consistency.

Tested by the first and fifth asserts in
`LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended.html?mouse`

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_boundary_events_attributes_during_drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseenter-mouseleave-on-drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=16-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=42-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&button=0&buttons=1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=16-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouseover-at-removing-mousedown-target_duration=42-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::swallowAnyClickEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):

Canonical link: https://commits.webkit.org/300273@main
